### ISSUE TITLE
Further enhance `new` syntax to reduce boilerplate

### ISSUE THREAD

**Issue by propensive** [lampepfl/dotty#5509]
I'm proposing this in the spirit of identifying "warts" with the language that could be cured, mainly to stimulate discussion.

In Scala 2.x, it is possible to write,
```
case class Foo[T]()
def foo: Foo[String] = new Foo()
```
and have the expected type `Foo[String]` able to infer the type parameter `T = String`.

Dotty goes further in allowing an anonymous class to be coerced into the expected type, e.g.,
```
trait Bar { def bar: Int }
val b: Bar = new { def bar = 42 }
```

But we could go even further along this road. In the presence of an expected type, we could invoke `new` eliding the type completely, for example,
```
class Baz(x: Int, y: String)
val baz: Baz = new(1, "two")
```
This could reduce the local boilerplate in many one-liners which assign a new instance to a value with an identical type. I searched my 50K lines of source code for, lines of the form,
```
... : X = new X ...
with
```
grep -r ': \(.*\) = new \(\1\)' *
```
and about 0.5% of the lines matched the pattern. So it happens quite frequently.

This would further be an opportunity to normalize the syntax of `new`. Currently, in a call such as,
```
new Foo().bar
```
the parser will read this as `(new Foo()).bar`. This is an anomaly in that the new (always followed by a space) has higher precedence than the `.` which comes after it.

The situation is further confused should the parentheses be omitted:
```
new Foo.bar
```
will attempt to instantiate a class called `bar` on the object `Foo`. Given that `new Foo` and `new Foo()` behave identically in isolation, this may be surprising, particularly to beginners looking for simple universal rules like "`.` always has higher precedence than ` `".

I therefore propose, for consistency with almost all other syntax in Scala, that `new` should accept a type, in square brackets immediately following it, which may be omitted if the type can be inferred from the expected type, just as with a method call, for example
```
class Foo(x: Int)
val f1: Foo = new(1)
val f2 = new[Foo](2)
val f3 = new(3) // error
```

In the cases where a type constructor is provided as a type parameter to `new`, instead of a proper type, this would be unified with the expected type to infer the type parameter, as currently happens (though generalized to type constructors, only really for consistency rather than utility), e.g.
```
class Foo[T]()
val f: Foo[String] = new[Foo]()
class Bar[A, B]()
val b: Bar[Int, String] = new[[T] => Bar[T, String]]()
```

All this would only work for class types, as is currently the case. Anonymous classes would still be written, `new {  ...  }`. Additionally, the existing `new` syntax should remain for the foreseeable future because the cost would drastically outweigh the benefits otherwise. Maybe it could have a 3-4 year window of deprecation, much like `val`s in for-comprehensions. But usage could certainly be opt-in to begin with.

**Comment by jducoeur** [comment]
My lizard-brain goes, "Ack!  Too different from other languages!  Scary!"

But it does fit Scala's "Consistency is Good" philosophy nicely -- stepping back and looking at it in the context of the rest of the language, the square brackets seem to make a lot of sense.

With the suggested timeline, I could see this one...

**Comment by lihaoyi** [comment]
C# is going to have this, they call it [target-typed new expressions](https://github.com/dotnet/csharplang/blob/master/proposals/target-typed-new.md):

```csharp
Dictionary<string, List<string>> musicians = new() {   { "John", new() { “guitar”, “bass” } }};
```

Would be nice to remove the boilerplate of defining implicit instances:

```scala
implicit val IntOrd: Ord[Int] = new {
  def compareTo(this x: Int)(y: Int) =
    if (x < y) -1 else if (x > y) +1 else 0
}
```

**Comment by propensive** [comment]
Yes, most of my examples were lines defining implicits, being the cases where I never omit the return type.

**Comment by propensive** [comment]
@arturopala Agreed. Though it's also only a tiny benefit, so I'd want to strike a good balance between forcing users to make changes they don't perceive to be particularly beneficial (automatic, or otherwise) and the good point you make about accommodating syntax variants. I don't remember the exact timings of the removal of `val` from `for`-comprehensions, but I think it was several years, but by the time it was removed, basically nobody was using it any more. :)

**Comment by hepin1989** [comment]
I am not sure if this is enabled now in Dotty, But I wondered this kind of object literal several times:
```scala
case class Person (name:String,age:Int)
val person = new {
 name = "me"
 age = 1
}
```

**Comment by nafg** [comment]
@propensive 
> val baz: Baz = new(1, "two")
Why is that better than
> val baz = new Baz(1, "two")
?

@lihaoyi 
> Would be nice to remove the boilerplate of defining implicit instances:
I'd rather if the requirement to explicitly type implicits were relaxed a bit. There are many cases where the type is known to be inferrable locally. For instance, if the right hand side is a simple `new` expression it should be permitted to rely on type inference. Similarly, `implicit object` should be permitted by the same logic.

@propensive 
> I therefore propose, for consistency with almost all other syntax in Scala, that new should accept a type, in square brackets immediately following it,

Why do we suddenly have a need to squeeze every good idea into dotty? Languages should evolve over time. Dotty should not be the once-in-a-lifetime opportunity to rethink Scala from the ground up. Unless this is supposed to be a new language. The goals of Dotty are
1. To have strong mathematical foundations
2. To clean up as many language warts as possible
3. To be a major version bump in the language
Since it is indeed a major version bump, we can indeed evolve the language, adding, changing, and removing features, but the same is true for all future major version bumps. Major version bumps shouldn't be so rare. C# is not much older than Scala and it's up to 8 already.

**Comment by jducoeur** [comment]
> Why is that better than
>> `val baz = new Baz(1, "two")`

It isn't necessarily -- but it's fairly common nowadays for company coding standards to require type ascriptions for all public members.  So in practice, boilerplatey duplication is common.

I agree that we should be getting more comfortable with language bumps: now that Scalafix is becoming real, they should start being less traumatic.  (Although we're going to learn a lot from the process of moving to Scala 3.)

That said, I think the proposal on the table -- which is, after all, suggesting that `new` is currently a language wart -- is worth considering for 3.0, especially if we do gradually transition.  We *can* do it later, so I don't think it's a crisis, but it looks like a smallish change compared to the rest...

**Comment by pavelfatin** [comment]
Here are a few related things from the IntelliJ Scala plugin that could be useful:
* [SCL-13575](https://youtrack.jetbrains.com/issue/SCL-13575) Type annotation settings: "Type is obvious" -> "Type is stable"
* [SCL-14339](https://youtrack.jetbrains.com/issue/SCL-14339) "definition type is obvious" specification

Ideally, we should not require a type annotation when the type is _stable_, and should not display a type _hint_ when the type is _obvious_.

**Comment by pavelfatin** [comment]
As for the `new` syntax - I would rather prefer `val baz = Baz(1, "two")` because:
1. Uniform access principle, for constructors.
2. Possible to toggle between constructor and factory method.
3. Consistency with case classes.
4. Constructor is a function.

**Comment by kjsingh** [comment]
+1 for present syntax. 

`val baz = Baz(1, "two")`

and

`val baz = (1, "two") // as tuple`

show clear intentions.

**Comment by sarveshseri** [comment]
I personally think that existence of "alternative" ways to write almost everything in Scala is brings more problems compared to benefits.

Yes, reduction of boilerplate is often good but it should not come at the cost of increased ambiguity, inconsistency and complexity.

**Comment by propensive** [comment]
Thank you to everyone who commented on this. It will take me a while to find the time, but I plan to write a SIP to propose this change for Scala 3. I will try to take into account (by incorporating, or addressing) all the feedback in the proposal.

**Comment by drdozer** [comment]
We need some fix for this boilerplate. For trivial things like `def foo: Foo = new Foo {}` it's a small win. For complicated things with type parameters and type members, it's very painful and quite error prone to include the type signature twice.

This is a simplified real code example I have that is one of the less complex cases:

```
implicit def logFilter[Log](implicit Log: LogDSL[Log]): Filter[Log] { E = Log } = new Filter[Log] {
  type E = Log

  def (e: Log) accept = "accept" gosub e
  def (e: Log) reject = "reject" gosub e
}
```

Note the type member being repeated in both the return type and the implementation. These cases become increasingly difficult to get right, particularly under refactoring, as the complicated return type has to correspond to multiple lines in the implementation.

**Comment by LPTK** [comment]
Allowing class constructors to be called with syntax `val baz = Baz(1, "two")` is never going to fly, as it would break virtually all existing code, and without an easy way to fix it. Indeed, we'd get ambiguities with `apply` methods declared in companion objects. Totally not worth it IMHO.

Moreover, it's also conceptually _wrong_. When you define a simple class `Baz` (not a case class) you only introduce a _type_ `Baz` into scope. Allowing syntax `Baz(1, "two")` makes it look like `Baz` is a value, but it's not – to see why, consider that `Baz.type` doesn't mean anything, and that `println(Baz)` won't work. So this change would blur the clean distinction that exists today between types and values in Scala

On the other hand, the `new[T]` syntax where the `[T]` can be inferred makes _a ton_ of sense. It would make the language feel more regular.

**Comment by drdozer** [comment]
FYI my issues with `new` syntax are largely obviated by the `instance of for` syntax variant.

**Comment by Atry** [comment]
`new` should be a method instead of a keyword.
I implemented similar feature in the type class [Factory](https://static.javadoc.io/com.thoughtworks.feature/the_2.11/2.3.0/com/thoughtworks/feature/Factory.html) in the library [feature.scala](https://github.com/ThoughtWorksInc/feature.scala).
